pageserver/src/layered_repository/timeline.rs

```diff
@@ -407,22 +372,82 @@ pub struct Timeline {
+    /// to modify this, it will also keep the prometheus metric in sync.
+    size_added_after_initial: AtomicI64,
+}
 
-    /// Information about the last processed message by the WAL receiver,
-    /// or None if WAL receiver has not received anything for this timeline
-    /// yet.
-    pub last_received_wal: Mutex<Option<WalReceiverInfo>>,
+/// Normalized current size, that the data in pageserver occupies.
+#[derive(Debug, Clone, Copy)]
+enum CurrentLogicalSize {
+    /// The size is not yet calculated to the end, this is an intermediate result,
+    /// constructed from walreceiver increments and normalized: logical data could delete some objects, hence be negative,
+    /// yet total logical size cannot be below 0.
+    Approximate(u64),
+    // Fully calculated logical size, only other future walreceiver increments are changing it, and those changes are
+    // available for observation without any calculations.
+    Exact(u64),
+}
 
-    /// Relation size cache
-    pub rel_size_cache: RwLock<HashMap<RelTag, (Lsn, BlockNumber)>>,
+impl CurrentLogicalSize {
+    fn size(&self) -> u64 {
+        *match self {
+            Self::Approximate(size) => size,
+            Self::Exact(size) => size,
+        }
+    }
+}
+
+impl LogicalSize {
+    fn empty_initial() -> Self {
+        Self {
+            initial_logical_size: OnceCell::with_value(0),
+            initial_part_end: None,
+            size_added_after_initial: AtomicI64::new(0),
+        }
+    }
+
+    fn deferred_initial(compute_to: Lsn) -> Self {
+        Self {
+            initial_logical_size: OnceCell::new(),
+            initial_part_end: Some(compute_to),
+            size_added_after_initial: AtomicI64::new(0),
+        }
+    }
+
+    fn current_size(&self) -> anyhow::Result<CurrentLogicalSize> {
+        let size_increment = self.size_added_after_initial.load(AtomicOrdering::Acquire);
+        match self.initial_logical_size.get() {
+            Some(initial_size) => {
+                let absolute_size_increment = u64::try_from(
+                    size_increment
+                        .checked_abs()
+                        .with_context(|| format!("Size added after initial {size_increment} is not expected to be i64::MIN"))?,
+                ).with_context(|| format!("Failed to convert size increment {size_increment} to u64"))?;
+
+                if size_increment < 0 {
+                    initial_size.checked_sub(absolute_size_increment)
+                } else {
+                    initial_size.checked_add(absolute_size_increment)
+                }.with_context(|| format!("Overflow during logical size calculation, initial_size: {initial_size}, size_increment: {size_increment}"))
+                .map(CurrentLogicalSize::Exact)
+            }
+            None => {
+                let non_negative_size_increment = size_increment.max(0);
+                u64::try_from(non_negative_size_increment)
+                    .with_context(|| {
```

Comment:
I'm just inconvenient with panics, that's why writing it this way, no good reason behind it, no strong opposition to other approach either, but would leave it as is either 🙂

Reply:
See https://rust.godbolt.org/z/cfY9xjWWP

Compiler managed to collapse the code and cut the possible panic. 

If you check with `cargo asm --asm-style=intel --rust pageserver::layered_repository::timeline::LogicalSize::current_size` there is no error message from this with_context anyway. So I think here we unnecessarily increase amount of code and if compiler stops seeing this behind layers of anyhow errors it will generate worse code. It is not on the hot code path, but anyway. Worth a comment at least. We can separate this to a separate function to decrease possibility of someone coming and changing something so panic becomes possible again. WDYT?

Reply:
Interesting, yes, we could do something better there indeed.

Reply:
This might be more clear as:
```
let non_negative_size_increment if size_increment < 0 {
    0
} else {
    size_increment as u64    
};
```
Or:
```
let non_negative_size_increment if size_increment < 0 {
    0
} else {
    u64::try_from(size_increment).unwrap("not negative, cannot fail")
};
```